src/lib.rs

```diff
@@ -346,10 +345,425 @@ impl From<Error> for io::Error {
+                    )
+                ),
+                path::Component::RootDir => (), // leading "/" does nothing
+                path::Component::ParentDir => self.truncate_to_parent()?,
+            }
+        }
+
+        Ok(())
+    }
+
+    fn truncate_to_parent(&mut self) -> Result<()> {
+        if self.pop() {
+            Ok(())
+        } else {
+            Err(Error::new(
+                io::Error::new(io::ErrorKind::NotFound, "path has no parent"),
+                "truncating to parent",
+                self.clone().into(),
+            ))
+        }
+    }
+
+    fn truncate_to_root(&mut self) {
+        let mut res = PathBuf::new();
+        for component in self.components().take(2) {
+            match component {
+                // We want to keep prefix and RootDir components of this path
+                | path::Component::Prefix(_)
+                | path::Component::RootDir
+                    => res.push(component),
+                // We want to discard all other components.
+                _ => break,
+            }
+        }
+
+        // Clobber ourselves with the new value.
+        *self = res;
+    }
+
+    fn set_file_name<S: AsRef<ffi::OsStr>>(&mut self, file_name: S) {
+        self.set_file_name(file_name)
+    }
+
+    fn set_extension<S: AsRef<ffi::OsStr>>(&mut self, extension: S) -> bool {
+        self.set_extension(extension)
+    }
+}
+
+impl PathMut for Arc<PathBuf> {
+    fn append<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
+        Arc::make_mut(self).append(path)
+    }
+    fn truncate_to_parent(&mut self) -> Result<()> {
+        Arc::make_mut(self).truncate_to_parent()
+    }
+    fn truncate_to_root(&mut self) {
+        Arc::make_mut(self).truncate_to_root()
+    }
+    fn set_file_name<S: AsRef<ffi::OsStr>>(&mut self, file_name: S) {
+        Arc::make_mut(self).set_file_name(file_name)
+    }
+    fn set_extension<S: AsRef<ffi::OsStr>>(&mut self, extension: S) -> bool {
+        Arc::make_mut(self).set_extension(extension)
+    }
+}
+
+/// Methods that return new path-like objects.
+///
+/// Like the methods of [`PathInfo`] and [`PathMut`], these methods are similar
+/// to ones from the standard library's [`PathBuf`] but may return a rich
+/// [`path_abs::Error`] instead of a [`std::io::Error`].
+///
+/// Unlike the methods of [`PathInfo`] and [`PathMut`], different types that
+/// implement this trait may have different return types.
+///
+/// [`PathInfo`]: trait.PathInfo.html
+/// [`PathMut`]: trait.PathInfo.html
+/// [`PathBuf`]: https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html
+/// [`path_abs::Error`]: struct.Error.html
+/// [`std::io::Error`]: https://doc.rust-lang.org/stable/std/io/struct.Error.html
+pub trait PathOps: PathInfo {
+    type Output: PathOps;
+
+    /// Returns a new value representing the concatenation of two paths.
+    ///
+    /// Note that this method represents pure concatenation, not "adjoinment"
```

Comment:
Please put a reference to what this means.

Reply:
Should be fixed in 9319001